libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -498,32 +541,69 @@ void AC_AttitudeControl::thrust_heading_rotation_angles(Quaternion& att_to_quat,
+    float desired_ang_vel = sqrt_controller(error_angle, smoothing_gain, accel_max, dt);
 
     // Acceleration is limited directly to smooth the beginning of the curve.
-    if (accel_max > 0) {
-        float delta_ang_vel = accel_max * _dt;
-        return constrain_float(ang_vel, target_ang_vel-delta_ang_vel, target_ang_vel+delta_ang_vel);
-    } else {
-        return ang_vel;
-    }
+    return input_shaping_ang_vel(target_ang_vel, desired_ang_vel, accel_max, dt);
 }
 
 // limits the acceleration and deceleration of a velocity request
-float AC_AttitudeControl::input_shaping_ang_vel(float target_ang_vel, float desired_ang_vel, float accel_max)
+float AC_AttitudeControl::input_shaping_ang_vel(float target_ang_vel, float desired_ang_vel, float accel_max, float dt)
 {
+    // Acceleration is limited directly to smooth the beginning of the curve.
     if (accel_max > 0.0f) {
-        float delta_ang_vel = accel_max * _dt;
-        target_ang_vel += constrain_float(desired_ang_vel - target_ang_vel, -delta_ang_vel, delta_ang_vel);
+        float delta_ang_vel = accel_max * dt;
+        return constrain_float(desired_ang_vel, target_ang_vel-delta_ang_vel, target_ang_vel+delta_ang_vel);
+    } else {
+        return desired_ang_vel;
+    }
+}
+
+// calculates the expected angular velocity correction from an angle error based on the AC_AttitudeControl settings.
+// This function can be used to predict the delay associated with angle requests.
+void AC_AttitudeControl::input_shaping_rate_predictor(Vector2f error_angle, Vector2f& target_ang_vel, float dt) const
+{
+    if (_rate_bf_ff_enabled & _use_ff_and_input_shaping) {
+        // translate the roll pitch and yaw acceleration limits to the euler axis
+        target_ang_vel.x = input_shaping_angle(wrap_PI(error_angle.x), _smoothing_gain, get_accel_roll_max_radss(), target_ang_vel.x, dt);
+        target_ang_vel.y = input_shaping_angle(wrap_PI(error_angle.y), _smoothing_gain, get_accel_pitch_max_radss(), target_ang_vel.y, dt);
+    } else {
+        target_ang_vel.x =  _p_angle_roll.get_p(wrap_PI(error_angle.x));
+        target_ang_vel.y =  _p_angle_pitch.get_p(wrap_PI(error_angle.y));
+    }
+    // Limit the angular velocity correction
+    Vector3f ang_vel(target_ang_vel.x, target_ang_vel.y, 0.0f);
+    ang_vel_limit(ang_vel, radians(_ang_vel_roll_max), radians(_ang_vel_pitch_max), 0.0f);
+
+    target_ang_vel.x =  ang_vel.x;
+    target_ang_vel.y =  ang_vel.y;
+}
+
+// limits angular velocity
+void AC_AttitudeControl::ang_vel_limit(Vector3f& euler_rad, float ang_vel_roll_max, float ang_vel_pitch_max, float ang_vel_yaw_max) const
+{
+    if (is_zero(ang_vel_pitch_max) || is_zero(ang_vel_pitch_max)) {
+        if (!is_zero(ang_vel_roll_max)) {
+            euler_rad.x = constrain_float(euler_rad.x, -ang_vel_roll_max, ang_vel_roll_max);
+        }
+        if (!is_zero(ang_vel_pitch_max)) {
+            euler_rad.y = constrain_float(euler_rad.y, -ang_vel_pitch_max, ang_vel_pitch_max);
+        }
```

Comment:
@lthall This nested if statement doesn't make sense.  first the outer if statement has a an or statement for the same test "is_zero(ang_vel_pitch_max)".  Then the nested if statement tests for the reverse of the outer if statement "!is_zero(ang_vel_pitch_max)".  Not say this is the issue that @ChristopherOlson found but found this inconsistency.

Reply:
The first statement checks to see if either one is zero the nested if statements then address if one of them is not zero.

It may have been more intuitive to check to see if both were not zero then handle that case then handle if either one is not zero and the other is.

Reply:
why does this first statement looking at pitch only in both evaluations and not pitch and roll
if (is_zero(ang_vel_pitch_max) || is_zero(ang_vel_pitch_max)) {

Reply:
Yep, dyslexic boy strikes again!!
That should be roll then pitch!!

Thanks Bill!

Reply:
@lthall Hi, so how about this PR? I'm trying to have a test but found that commit [Copter: Loiter updates](https://github.com/lthall/Leonard_ardupilot/tree/a1e52acdc0586ea77e96aaeae22fc52954408315) has disabled this function?

Reply:
@roger-zhao Can you please explain what you are asking

Reply:
@lthall Sorry for my unclear description, what I'm saying it that when I try to build a FW from source code in Randy's branch [lthall-new-loiter5](https://github.com/rmackay9/rmackay9-ardupilot/tree/lthall-new-loiter5), I've found that branch is different with source code in your branch [NewLoiter-09122017](https://github.com/lthall/Leonard_ardupilot/tree/NewLoiter-09122017) , and after source code comparing, it seems like  in your branch, the new-loiter's "mapping  pilot input to lean-angle implementation" has been deleted again (mightbe I'm wrong with this becasue I'm just doing a source code comparing), so I'm not sure if I should build FW from Randy's branch [lthall-new-loiter5](https://github.com/rmackay9/rmackay9-ardupilot/tree/lthall-new-loiter5) or  your branch [NewLoiter-09122017](https://github.com/lthall/Leonard_ardupilot/tree/NewLoiter-09122017) for testing this new loiter.